In [1]:
import pickle
import re
import string
import nltk
import pandas as pd
import spacy
import re
import string
from collections import Counter

# stemmerl = LancasterStemmer()
# stemmerp = PorterStemmer()
# stemmers = SnowballStemmer('english')
# stemmerwn = WordNetLemmatizer()
# spacy.prefer_gpu()

In [2]:
df = pd.read_pickle('cleaned_list_df.pickle')
t = df.transpose()

In [3]:
t

wsb  \
0       in the future when apple is at  a share your g...   
1       you ever stare at your account and think wow h...   
2                                  buy  at the top gang 🤡   
3       almost responded to an email with thanks retar...   
4       top trending tickers sentiments and options on...   
...                                                   ...   
101136                                               none   
101137                                               none   
101138                                               none   
101139                                               none   
101140                                               none   

                                                  science  \
0        the children were not motivated by achieving ...   
1       also proved children like to gamble according ...   
2       kind of like how gamers go through rpg games e...   
3       i teach elementary pe  to  year olds one of th...   
4       wouldnt this make lootmystery boxes even worse...   
...                                                   ...   
101136                                               none   
101137                                               none   
101138                                               none   
101139                                               none   
101140                                               none   

                                                      ama  \
0                      tldr do whatever the fuck you want   
1          so youre telling me people on the internet lie   
2                                                           
3       this is so funny this subreddit got fucked by ...   
4       a note about celebrity request bans when riama...   
...                                                   ...   
101136                                               none   
101137                                               none   
101138                                               none   
101139                                               none   
101140                                               none   

                                                askreddit  \
0       victorias secret was originally supposed to be...   
1       the ford mustang  it was supposed to be a car ...   
2                           heels if i remember correctly   
3       is this how family feud gets new questions and...   
4       skirts used to be worn in battle by men for mo...   
...                                                   ...   
101136                                              maybe   
101137  is it my cake day i have always wondered how p...   
101138  theres a cake emoji next to your username on y...   
101139  i think my problem is that im using narwhal ed...   
101140                                                lol   

                                                worldnews  \
0             so was there any posts left after the purge   
1       ok just for some context  facebook has an esti...   
2       they still havent taken down my post about mar...   
3       why does facebook have a fake news problem  be...   
4       a lot of karens are going to be be saying they...   
...                                                   ...   
101136                                               none   
101137                                               none   
101138                                               none   
101139                                               none   
101140                                               none   

                                                    funny  \
0                           but i wanted the robot ferret   
1       haha suck it you fools i already got my roboti...   
2         this not funny rule number oh i forget carry on   
3                     but the nigerian guy is legit right   
4       bruh when i was younger i went to a farm this ...   
...                    

In [5]:
# Clean text

def clean_text(text):
    '''Make text lowercase, remove text in square brackets, remove punctuation, remove words containing numbers, remove newlines.'''
    text = text.lower()
    text = re.sub('\[.*?\]', '', text)
    text = re.sub('[%s]' % re.escape(string.punctuation), '', text)
    text = re.sub('\w*\d\w*', '', text)
    text = re.sub('[‘’“”…]', '', text)
    text = re.sub('\n', ' ', text)
    return text

clean = lambda x: clean_text(x)

# # Let's take a look at the updated text
# data_clean = pd.DataFrame(t.text.apply(clean))
# data_clean['subreddit'] = subs
# data_clean

In [10]:
doc = ['this is a test document',
      'pretty cool huh?',
      'you are lemman']

In [11]:
def lemmatize(text):
    nlp = spacy.load('en_core_web_sm', disable=['parser', 'ner'])
    nlp.max_length = 14082871
    lemmed = []
    for comment in text:
        teckst = nlp(comment)
        lemmed.append(' '.join([word.lemma_ for word in teckst if not word.is_stop]))
    return lemmed

In [14]:
ar = t['askreddit'].tolist()
ar

['victorias secret was originally supposed to be a place for men to buy lingerie for their wives and not feel embarrassed but now its almost exclusively shopped at by women',
 'the ford mustang  it was supposed to be a car with a big trunk to you could fit all the groceries in it  there was even an ad or article stating that you can own a mustang that matches the color of your lipstick  makes sense why a lot of old mustangs are red',
 'heels if i remember correctly',
 'is this how family feud gets new questions and answers',
 'skirts used to be worn in battle by men for more room to move around',
 'ugg boots were originally worn by male surfers with their girlfriends stealing and wearing them like they would a leather jacket',
 'filtered cigarettes in general but marlboro made the biggest leap in terms of gendered marketing once that whole conspiracy theory came out that like cigarettes are idk bad for you or something and more people switched to filtered cigarettes its filtered that m

In [15]:
lemd = lemmatize(ar)
lemd

KeyboardInterrupt: 

In [36]:
df = pd.DataFrame(lemd)
df.to_pickle('askreddit_lemmatized.pickle')

In [55]:
data_clean.to_pickle("data_clean.pickle")

In [56]:
import numpy as np
arr = []
for i, text in enumerate(data_clean.text):
    arr.append(len(data_clean.text[i]))
    
lengths = np.asarray(arr).sum() 
print(lengths)
per_gb = lengths / 100000
per_gb

47057853


470.57853

In [37]:
df = pd.read_pickle('askreddit_lemmatized.pickle')
df

0
0                                 victorias secret originally suppose place man buy lingerie wife feel embarrassed exclusively shop woman
1       ford mustang   suppose car big trunk fit grocery   ad article state mustang match color lipstick   make sense lot old mustang red
2                                                                                                                 heel remember correctly
3                                                                                                     family feud get new question answer
4                                                                                                              skirt wear battle man room
...                                                                                                                                   ...
101136                                                                                                                              maybe
101137                                                                                                        cake day wonder people know
101138                                                     s cake emoji username comment cake day check profile cake day   happy cake day
101139                          think problem be narwhal edit thank sorry get distract excitement find cake day think post picture kitten
101140                                                                                                                                lol

[101141 rows x 1 columns]

In [57]:
cv = CountVectorizer(stop_words='english')
data_cv = cv.fit_transform(data_clean.text)
dtm_cv = pd.DataFrame(data_cv.toarray(), columns=cv.get_feature_names())
dtm_cv.index = data_clean.index
dtm_cv

aa  aaa  aaaaa  aaaaaaaa  aaaaaaaaaaaa  aaaaaaaaaaaaaaaa  \
wsb                 0    5      0         0             0                 0   
science            10    0      0         0             0                 0   
ama                 0    0      0         0             0                 0   
askreddit           9    3      0         1             1                 0   
worldnews           2    0      0         0             1                 0   
funny               0    0      0         0             0                 1   
dankmemes           0    0      0         0             0                 0   
memes               0    0      1         0             0                 0   
nosleep             1    0      0         0             0                 0   
psychology          2    0      0         0             0                 0   
politics            2    2      0         0             0                 0   
cscareerquestions   0    0      0         0             0                 0   
writingprompts      0    0      0         0             0                 0   

                   aaaaaaaaaaaaaaaaaaaaaa  aaaaaaaaaaaaaaaaaaaaaaaa  \
wsb                                     1                         0   
science                                 0                         0   
ama                                     0                         0   
askreddit                               0                         1   
worldnews                               0                         0   
funny                                   0                         0   
dankmemes                               0                         0   
memes                                   0                         0   
nosleep                                 0                         0   
psychology                              0                         0   
politics                                0                         0   
cscareerquestions                       0                         0   
writingprompts                          0                         0   

                   aaaaaaaaaaaaaaaaaaaaaaaaaaaa  \
wsb                                           0   
science                                       0   
ama                                           0   
askreddit                                     0   
worldnews                                     0   
funny                                         0   
dankmemes                                     0   
memes                                         8   
nosleep                                       0   
psychology                                    0   
politics                                      0   
cscareerquestions                             0   
writingprompts                                0   

                   aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa  ...  𝘽𝙊𝘽  𝘽𝙐𝙄𝙇𝘿𝙀𝙍  𝘾𝘼𝙉  \
wsb                                                0  ...    0        0    0   
science                                            0  ...    0        0    0   
ama                                                0  ...    0        0    0   
askreddit                                          1  ...    2        2    2   
worldnews                                          0  ...    0        0    0   
funny                                              0  ...    0        0    0   
dankmemes                                          0  ...    0        0    0   
memes                                              0  ...    0        0    0   
nosleep                                            0  ...    0        0    0   
psychology                                         0  ...    0        0    0   
politics                                           0  ...    0        0    0   
cscareerquestions                                  0  ...    0        0    0   
writingprompts                                     0  ...    0        0    0   

                   𝙁𝙄𝙓  𝙃𝙀  𝙄𝙏  𝙏𝙃𝙀  𝙔𝙀𝙎  𝚃𝚑𝚒𝚜  𝚝𝚎𝚖𝚙𝚕𝚊𝚝𝚎  
wsb                  0

In [58]:
tf_idf = TfidfVectorizer(stop_words='english')
data_tf = tf_idf.fit_transform(data_clean.text)
dtm_idf = pd.DataFrame(data_tf.toarray(), columns=tf_idf.get_feature_names())
dtm_idf.index = data_clean.index
dtm_idf

aa       aaa     aaaaa  aaaaaaaa  aaaaaaaaaaaa  \
wsb                0.000000  0.001008  0.000000  0.000000      0.000000   
science            0.000551  0.000000  0.000000  0.000000      0.000000   
ama                0.000000  0.000000  0.000000  0.000000      0.000000   
askreddit          0.000375  0.000166  0.000000  0.000073      0.000063   
worldnews          0.000421  0.000000  0.000000  0.000000      0.000316   
funny              0.000000  0.000000  0.000000  0.000000      0.000000   
dankmemes          0.000000  0.000000  0.000000  0.000000      0.000000   
memes              0.000000  0.000000  0.001941  0.000000      0.000000   
nosleep            0.000498  0.000000  0.000000  0.000000      0.000000   
psychology         0.003142  0.000000  0.000000  0.000000      0.000000   
politics           0.000171  0.000227  0.000000  0.000000      0.000000   
cscareerquestions  0.000000  0.000000  0.000000  0.000000      0.000000   
writingprompts     0.000000  0.000000  0.000000  0.000000      0.000000   

                   aaaaaaaaaaaaaaaa  aaaaaaaaaaaaaaaaaaaaaa  \
wsb                        0.000000                0.000264   
science                    0.000000                0.000000   
ama                        0.000000                0.000000   
askreddit                  0.000000                0.000000   
worldnews                  0.000000                0.000000   
funny                      0.001627                0.000000   
dankmemes                  0.000000                0.000000   
memes                      0.000000                0.000000   
nosleep                    0.000000                0.000000   
psychology                 0.000000                0.000000   
politics                   0.000000                0.000000   
cscareerquestions          0.000000                0.000000   
writingprompts             0.000000                0.000000   

                   aaaaaaaaaaaaaaaaaaaaaaaa  aaaaaaaaaaaaaaaaaaaaaaaaaaaa  \
wsb                                0.000000                      0.000000   
science                            0.000000                      0.000000   
ama                                0.000000                      0.000000   
askreddit                          0.000073                      0.000000   
worldnews                          0.000000                      0.000000   
funny                              0.000000                      0.000000   
dankmemes                          0.000000                      0.000000   
memes                              0.000000                      0.015531   
nosleep                            0.000000                      0.000000   
psychology                         0.000000                      0.000000   
politics                           0.000000                      0.000000   
cscareerquestions                  0.000000                      0.000000   
writingprompts                     0.000000                      0.000000   

                   aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa  ...       𝘽𝙊𝘽   𝘽𝙐𝙄𝙇𝘿𝙀𝙍  \
wsb                                         0.000000  ...  0.000000  0.000000   
science                                     0.000000  ...  0.000000  0.000000   
ama                                         0.000000  ...  0.000000  0.000000   
askreddit                                   0.000073  ...  0.000145  0.000145   
worldnews                                   0.000000  ...  0.000000  0.000000   
funny                                       0.000000  ...  0.000000  0.000000   
dankmemes                                   0.000000  ...  0.000000  0.000000   
memes                                       0.000000  ...  0.000000  0.000000   
nosleep                                     0.000000  ...  0.000000  0.000000   
psychology                                  0.000000  ...  0.000000  0.000000   
politics                                    0.000000  ...  0.000000  0.000000   
cscareerquestions                       

In [59]:
# Pickling document-term-matrices
dtm_cv.to_pickle("dtm_cv.pickle")
dtm_idf.to_pickle("dtm_idf.pickle")

# Pickling cleaned dataframe
data_clean.to_pickle('data_clean.pickle')
pickle.dump(cv, open("count_vectorizer.pickle", "wb"))
pickle.dump(cv, open("tf_idf.pickle", "wb"))

In [60]:
# Read in the document-term matrix

data_cv = pd.read_pickle('dtm_cv.pickle')
data_cv = data_cv.transpose()
data_cv

wsb  science  ama  askreddit  worldnews  funny  dankmemes  \
aa              0       10    0          9          2      0          0   
aaa             5        0    0          3          0      0          0   
aaaaa           0        0    0          0          0      0          0   
aaaaaaaa        0        0    0          1          0      0          0   
aaaaaaaaaaaa    0        0    0          1          1      0          0   
...           ...      ...  ...        ...        ...    ...        ...   
𝙄𝙏              0        0    0          1          0      0          0   
𝙏𝙃𝙀             0        0    0          2          0      0          0   
𝙔𝙀𝙎             0        0    0          1          0      0          0   
𝚃𝚑𝚒𝚜            0        0    0          0          0      0          0   
𝚝𝚎𝚖𝚙𝚕𝚊𝚝𝚎        0        0    0          0          0      0          0   

              memes  nosleep  psychology  politics  cscareerquestions  \
aa                0        1           2         2                  0   
aaa               0        0           0         2                  0   
aaaaa             1        0           0         0                  0   
aaaaaaaa          0        0           0         0                  0   
aaaaaaaaaaaa      0        0           0         0                  0   
...             ...      ...         ...       ...                ...   
𝙄𝙏                0        0           0         0                  0   
𝙏𝙃𝙀               0        0           0         0                  0   
𝙔𝙀𝙎               0        0           0         0                  0   
𝚃𝚑𝚒𝚜              1        0           0         0                  0   
𝚝𝚎𝚖𝚙𝚕𝚊𝚝𝚎          1        0           0         0                  0   

              writingprompts  
aa                         0  
aaa                        0  
aaaaa                      0  
aaaaaaaa                   0  
aaaaaaaaaaaa               0  
...                      ...  
𝙄𝙏                         0  
𝙏𝙃𝙀                        0  
𝙔𝙀𝙎                        0  
𝚃𝚑𝚒𝚜                       0  
𝚝𝚎𝚖𝚙𝚕𝚊𝚝𝚎                   0  

[111923 rows x 13 columns]

In [61]:
data_idf = pd.read_pickle('dtm_idf.pickle')
data_idf = data_idf.transpose()
data_idf

wsb   science  ama  askreddit  worldnews  funny  dankmemes  \
aa            0.000000  0.000551  0.0   0.000375   0.000421    0.0        0.0   
aaa           0.001008  0.000000  0.0   0.000166   0.000000    0.0        0.0   
aaaaa         0.000000  0.000000  0.0   0.000000   0.000000    0.0        0.0   
aaaaaaaa      0.000000  0.000000  0.0   0.000073   0.000000    0.0        0.0   
aaaaaaaaaaaa  0.000000  0.000000  0.0   0.000063   0.000316    0.0        0.0   
...                ...       ...  ...        ...        ...    ...        ...   
𝙄𝙏            0.000000  0.000000  0.0   0.000073   0.000000    0.0        0.0   
𝙏𝙃𝙀           0.000000  0.000000  0.0   0.000145   0.000000    0.0        0.0   
𝙔𝙀𝙎           0.000000  0.000000  0.0   0.000073   0.000000    0.0        0.0   
𝚃𝚑𝚒𝚜          0.000000  0.000000  0.0   0.000000   0.000000    0.0        0.0   
𝚝𝚎𝚖𝚙𝚕𝚊𝚝𝚎      0.000000  0.000000  0.0   0.000000   0.000000    0.0        0.0   

                 memes   nosleep  psychology  politics  cscareerquestions  \
aa            0.000000  0.000498    0.003142  0.000171                0.0   
aaa           0.000000  0.000000    0.000000  0.000227                0.0   
aaaaa         0.001941  0.000000    0.000000  0.000000                0.0   
aaaaaaaa      0.000000  0.000000    0.000000  0.000000                0.0   
aaaaaaaaaaaa  0.000000  0.000000    0.000000  0.000000                0.0   
...                ...       ...         ...       ...                ...   
𝙄𝙏            0.000000  0.000000    0.000000  0.000000                0.0   
𝙏𝙃𝙀           0.000000  0.000000    0.000000  0.000000                0.0   
𝙔𝙀𝙎           0.000000  0.000000    0.000000  0.000000                0.0   
𝚃𝚑𝚒𝚜          0.001941  0.000000    0.000000  0.000000                0.0   
𝚝𝚎𝚖𝚙𝚕𝚊𝚝𝚎      0.001941  0.000000    0.000000  0.000000                0.0   

              writingprompts  
aa                       0.0  
aaa                      0.0  
aaaaa                    0.0  
aaaaaaaa                 0.0  
aaaaaaaaaaaa             0.0  
...                      ...  
𝙄𝙏                       0.0  
𝙏𝙃𝙀                      0.0  
𝙔𝙀𝙎                      0.0  
𝚃𝚑𝚒𝚜                     0.0  
𝚝𝚎𝚖𝚙𝚕𝚊𝚝𝚎                 0.0  

[111923 rows x 13 columns]

In [35]:
# Find the top 30 words in each subreddit
top_dict_cv = {}
for c in data_cv.columns:
    top = data_cv[c].sort_values(ascending=False).head(30)
    top_dict_cv[c]= list(zip(top.index, top.values))

top_dict_cv

{'wsb': [('just', 2801),
  ('im', 2434),
  ('like', 2322),
  ('calls', 2270),
  ('dont', 1652),
  ('buy', 1650),
  ('tsla', 1555),
  ('today', 1420),
  ('going', 1381),
  ('money', 1320),
  ('day', 1286),
  ('puts', 1234),
  ('market', 1220),
  ('corn', 1210),
  ('good', 1194),
  ('bought', 1163),
  ('lol', 1160),
  ('spy', 1129),
  ('fuck', 1127),
  ('think', 1126),
  ('sell', 1126),
  ('shit', 1096),
  ('stock', 1082),
  ('time', 1071),
  ('got', 1014),
  ('people', 954),
  ('fucking', 925),
  ('right', 918),
  ('tesla', 883),
  ('thats', 863)],
 'science': [('people', 11366),
  ('just', 8939),
  ('like', 8171),
  ('dont', 7114),
  ('im', 5866),
  ('think', 4672),
  ('know', 3985),
  ('thats', 3966),
  ('time', 3796),
  ('really', 3451),
  ('make', 3370),
  ('way', 3317),
  ('good', 3059),
  ('need', 2976),
  ('years', 2861),
  ('work', 2770),
  ('doesnt', 2657),
  ('youre', 2596),
  ('going', 2591),
  ('things', 2454),
  ('lot', 2431),
  ('use', 2414),
  ('isnt', 2282),
  ('does', 2

In [36]:
# Find the top 30 words in each subreddit
top_dict_idf = {}
for c in data_idf.columns:
    top = data_idf[c].sort_values(ascending=False).head(30)
    top_dict_idf[c]= list(zip(top.index, top.values))

top_dict_idf

{'wsb': [('tsla', 0.31347728036514566),
  ('just', 0.2506532195205123),
  ('im', 0.21781147315706068),
  ('like', 0.20778892385813266),
  ('calls', 0.20313559739791606),
  ('slv', 0.16845369176551078),
  ('dont', 0.14783260215918825),
  ('buy', 0.14765362806456456),
  ('spy', 0.13502496137027253),
  ('today', 0.12707160718283736),
  ('corn', 0.12497065915376408),
  ('going', 0.12358161233767494),
  ('puts', 0.1186105386379494),
  ('money', 0.11812290245165163),
  ('day', 0.1150803428430485),
  ('tesla', 0.11392938909985918),
  ('stock', 0.11175062248295269),
  ('aapl', 0.11139509353048128),
  ('market', 0.10917419772046591),
  ('good', 0.10684753449035761),
  ('earnings', 0.10469668662021958),
  ('bought', 0.10407343602369004),
  ('lol', 0.10380497488175447),
  ('fuck', 0.10085190232046318),
  ('sell', 0.10076241527315132),
  ('think', 0.10076241527315132),
  ('shit', 0.0980778038537956),
  ('time', 0.09584062767099917),
  ('got', 0.0907398659742233),
  ('rkt', 0.0877857266947028)],
 '

In [37]:
# Print the top 15 words in each subreddit
for subreddit, top_words in top_dict_cv.items():
    print(subreddit)
    print(', '.join([word for word, count in top_words[0:30]]))
    print('---')

wsb
just, im, like, calls, dont, buy, tsla, today, going, money, day, puts, market, corn, good, bought, lol, spy, fuck, think, sell, shit, stock, time, got, people, fucking, right, tesla, thats
---
science
people, just, like, dont, im, think, know, thats, time, really, make, way, good, need, years, work, doesnt, youre, going, things, lot, use, isnt, does, say, study, right, better, actually, want
---
ama
im, like, did, just, dont, think, people, time, thats, happy, birthday, ive, know, want, good, say, lol, really, money, day, years, thanks, question, youre, lot, got, feel, thing, make, hope
---
askreddit
just, like, im, dont, people, time, know, think, good, really, years, got, didnt, ive, make, thats, want, did, life, way, sex, day, feel, right, going, youre, things, say, better, said
---
worldnews
people, just, like, dont, im, think, thats, russia, country, china, know, world, time, mask, going, right, really, youre, government, nuclear, way, good, new, make, vaccine, war, doesnt, w

In [38]:
# Print the top 15 words in each subreddit
for subreddit, top_words in top_dict_idf.items():
    print(subreddit)
    print(', '.join([word for word, count in top_words[0:30]]))
    print('---')

wsb
tsla, just, im, like, calls, slv, dont, buy, spy, today, corn, going, puts, money, day, tesla, stock, aapl, market, good, earnings, bought, lol, fuck, sell, think, shit, time, got, rkt
---
science
people, just, like, dont, im, think, know, thats, time, really, make, way, good, need, years, work, doesnt, youre, going, things, lot, use, isnt, does, say, study, right, better, actually, want
---
ama
im, like, did, just, dont, think, people, time, birthday, thats, happy, ive, know, want, good, lol, say, really, alters, money, day, years, thanks, question, youre, got, lot, thing, feel, make
---
askreddit
just, like, im, dont, people, time, know, think, good, really, years, got, didnt, ive, make, thats, want, did, life, way, sex, day, feel, right, going, youre, things, say, better, said
---
worldnews
people, just, like, dont, im, russia, think, thats, country, china, know, world, nuclear, time, mask, going, vaccine, government, nz, right, really, youre, way, good, new, make, war, virus, d

Significantly different results using CountVectorizer and TF-IDF for vectorization.

Pretty interesting how tipping /gratuity / restaurants are one of the top topics of r/science. I guess tipping is a hot subject amongst redditors. Also, autism appears in both searches. Autism appeared in two forms: autistic and autism. Maybe stemming will consolidate these. Let's try it!

In [39]:
words = []
for subreddit in data_cv.columns:
    top = [word for (word, count) in top_dict_cv[subreddit]]
    for t in top:
        words.append(t)

words

Counter(words).most_common()

[('just', 13),
 ('im', 13),
 ('like', 13),
 ('dont', 13),
 ('time', 13),
 ('good', 12),
 ('think', 12),
 ('people', 12),
 ('thats', 12),
 ('know', 12),
 ('make', 11),
 ('really', 10),
 ('youre', 10),
 ('way', 9),
 ('right', 8),
 ('say', 7),
 ('want', 7),
 ('going', 6),
 ('got', 6),
 ('did', 6),
 ('doesnt', 5),
 ('ive', 5),
 ('day', 4),
 ('lol', 4),
 ('need', 4),
 ('years', 4),
 ('things', 4),
 ('lot', 4),
 ('fuck', 3),
 ('work', 3),
 ('does', 3),
 ('thing', 3),
 ('said', 3),
 ('new', 3),
 ('money', 2),
 ('sell', 2),
 ('shit', 2),
 ('use', 2),
 ('study', 2),
 ('better', 2),
 ('actually', 2),
 ('feel', 2),
 ('hope', 2),
 ('didnt', 2),
 ('life', 2),
 ('sure', 2),
 ('man', 2),
 ('meme', 2),
 ('yes', 2),
 ('looks', 2),
 ('story', 2),
 ('calls', 1),
 ('buy', 1),
 ('tsla', 1),
 ('today', 1),
 ('puts', 1),
 ('market', 1),
 ('corn', 1),
 ('bought', 1),
 ('spy', 1),
 ('stock', 1),
 ('fucking', 1),
 ('tesla', 1),
 ('isnt', 1),
 ('happy', 1),
 ('birthday', 1),
 ('thanks', 1),
 ('question', 1),
 ('

In [40]:
# Most commonly used words and how many subreddits they appear in (we're looking at a total of 12 subreddits)
Counter(words).most_common()

[('just', 13),
 ('im', 13),
 ('like', 13),
 ('dont', 13),
 ('time', 13),
 ('good', 12),
 ('think', 12),
 ('people', 12),
 ('thats', 12),
 ('know', 12),
 ('make', 11),
 ('really', 10),
 ('youre', 10),
 ('way', 9),
 ('right', 8),
 ('say', 7),
 ('want', 7),
 ('going', 6),
 ('got', 6),
 ('did', 6),
 ('doesnt', 5),
 ('ive', 5),
 ('day', 4),
 ('lol', 4),
 ('need', 4),
 ('years', 4),
 ('things', 4),
 ('lot', 4),
 ('fuck', 3),
 ('work', 3),
 ('does', 3),
 ('thing', 3),
 ('said', 3),
 ('new', 3),
 ('money', 2),
 ('sell', 2),
 ('shit', 2),
 ('use', 2),
 ('study', 2),
 ('better', 2),
 ('actually', 2),
 ('feel', 2),
 ('hope', 2),
 ('didnt', 2),
 ('life', 2),
 ('sure', 2),
 ('man', 2),
 ('meme', 2),
 ('yes', 2),
 ('looks', 2),
 ('story', 2),
 ('calls', 1),
 ('buy', 1),
 ('tsla', 1),
 ('today', 1),
 ('puts', 1),
 ('market', 1),
 ('corn', 1),
 ('bought', 1),
 ('spy', 1),
 ('stock', 1),
 ('fucking', 1),
 ('tesla', 1),
 ('isnt', 1),
 ('happy', 1),
 ('birthday', 1),
 ('thanks', 1),
 ('question', 1),
 ('

In [41]:
add_stop_words = [word for word, count in Counter(words).most_common() if count > 4]
add_stop_words

['just',
 'im',
 'like',
 'dont',
 'time',
 'good',
 'think',
 'people',
 'thats',
 'know',
 'make',
 'really',
 'youre',
 'way',
 'right',
 'say',
 'want',
 'going',
 'got',
 'did',
 'doesnt',
 'ive']

In [42]:
# Add new stop words
stop_words = text.ENGLISH_STOP_WORDS.union(add_stop_words)

# Recreate document-term matrix
cv = CountVectorizer(stop_words=stop_words)
data_cv = cv.fit_transform(data_clean.text)
data_stop = pd.DataFrame(data_cv.toarray(), columns=cv.get_feature_names())
data_stop.index = data_clean.index

pickle.dump(cv, open("cv_stop.pickle", "wb"))
data_stop.to_pickle("dtm_stop.pickle")

AttributeError: 'str' object has no attribute 'ENGLISH_STOP_WORDS'

In [62]:
from wordcloud import WordCloud

wc = WordCloud(background_color='white', colormap='Dark2', max_font_size=100, random_state=42)

In [63]:
import matplotlib.pyplot as plt

plt.rcParams['figure.figsize'] = [20, 20]



for index, subreddit in enumerate(dtm_cv.columns):
    wc.generate(data_clean.text[subreddit])
    
    plt.subplot(4, 3, index + 1)
    plt.imshow(wc, interpolation='bilinear')
    plt.axis('off')
    plt.title(subreddit)

plt.show()

KeyError: 'aa'

Oof, these look terrible. I'm following Alice's NLP tutorial and I like the succinct way she wrote a reusable piece of code to put all comments in a dictionary as values, with the subreddit name as the key, then turning each subreddit's list of comments into a single giant string. However, when I try to stem/lemmatize these huge strings after cleaning them, the computer reaches its recursive limit. So I may need to go back and enter each corpus in manually (as I do a bit further below), then clean and lemmatize before creating the large strings. 

That will allow me to lemmatize and stem - however, removing stop words is probably the best way to improve the word clouds. (Also, they already look stemmed --- hmm) It also looks like there are bi-grams in the word clouds, even though I never specified bi-grams. Does WordCloud do these things in-house?

In [64]:
# Find the number of unique words used in each subreddit

# Identify the non-zero items in the document-term matrix, meaning that the word occurs at least once
unique_list = []
dtm_cv = dtm_cv.transpose()
for sub in dtm_cv.columns:
    uniques = dtm_cv[sub].to_numpy().nonzero()[0].size
    unique_list.append(uniques)

# Create a new dataframe that contains this unique word count
data_words = pd.DataFrame(list(zip(subs, unique_list)), columns=['subreddit', 'unique_words'])
data_unique_sort = data_words.sort_values(by='unique_words', ascending=False)
data_unique_sort

subreddit  unique_words
3           askreddit         54730
1             science         47393
10           politics         31623
12     writingprompts         25989
4           worldnews         21091
11  cscareerquestions         20052
0                 wsb         19380
8             nosleep         10870
5               funny          9956
7               memes          8110
9          psychology          7410
6           dankmemes          6267
2                 ama          4587

r/science has the most unique words - no surprise there. 

**this was an error -->** But r/memes has the second highest words?? Now that is a surprise. Maybe they use lots of random, non-real words as a kind of 'baby talk' or altering normal words? 

It's interesting to me that wallstreetbets has the least unique words, because they have lots of lingo and there seems to be a variety of different comments on there. I guess the lingo can work against the unique word count, too - they keep using the same words over and over. 

Finally, note that nosleep is 3rd from last in the unique word count. That makes sense - I was thinking at first that this included post submissions (not just comments) and that the low unique word count reflects the cliched and overused writing (creepy stories). However, my analysis is just of comments, not of submissions. So the comments are even less unique - probably just repackaged versions of, 'wow, spooky story! i'm afraid to turn the lights off now' etc.

In [ ]:
with open(f'/home/nick/Documents/Reddit_NLP/pickles/wsb.pickle', 'rb') as file:
    wsb = pickle.load(file)
with open('/home/nick/Documents/Reddit_NLP/pickles/science.pickle', 'rb') as file:
    science = pickle.load(file)
with open('/home/nick/Documents/Reddit_NLP/pickles/ama.pickle', 'rb') as file:
    ama = pickle.load(file)
with open('/home/nick/Documents/Reddit_NLP/pickles/askreddit.pickle', 'rb') as file:
    askreddit = pickle.load(file)
with open('/home/nick/Documents/Reddit_NLP/pickles/worldnews.pickle', 'rb') as file:
    worldnews = pickle.load(file)
with open('/home/nick/Documents/Reddit_NLP/pickles/funny.pickle', 'rb') as file:
    funny = pickle.load(file)
with open('/home/nick/Documents/Reddit_NLP/pickles/dankmemes.pickle', 'rb') as file:
    dankmemes = pickle.load(file)
with open('/home/nick/Documents/Reddit_NLP/pickles/politics.pickle', 'rb') as file:
    politics = pickle.load(file)
with open('/home/nick/Documents/Reddit_NLP/pickles/memes.pickle', 'rb') as file:
    memes = pickle.load(file)
with open('/home/nick/Documents/Reddit_NLP/pickles/psychology.pickle', 'rb') as file:
    psychology = pickle.load(file)
with open('/home/nick/Documents/Reddit_NLP/pickles/nosleep.pickle', 'rb') as file:
    nosleep = pickle.load(file)
with open('/home/nick/Documents/Reddit_NLP/pickles/cscareerquestions.pickle', 'rb') as file:
    cscareerquestions = pickle.load(file)

In [ ]:
print('wsb: {}'.format(len(wsb)))
print('science: {}'.format(len(science)))
print('ama: {}'.format(len(ama)))
print('askreddit: {}'.format(len(askreddit)))    
print('worldnews: {}'.format(len(worldnews)))      
print('funny: {}'.format(len(funny)))      
print('dankmemes: {}'.format(len(dankmemes)))      
print('politics: {}'.format(len(politics)))      
print('psychology: {}'.format(len(psychology)))   
print('cscareerquestions: {}'.format(len(cscareerquestions)))   
print('nosleep: {}'.format(len(nosleep)))   
print('memes: {}'.format(len(memes)))   

In [78]:
# Calculate the words per comment of each subreddit

# Find the total number of words that a subreddit uses
total_list = []
sub_name = []
for subreddit in dtm_cv.columns:
    totals = sum(dtm_cv[subreddit])
    sub_name.append(subreddit)
    total_list.append(totals)
    
data_words = pd.DataFrame(list(zip(dtm_cv.columns, total_list)), columns=['subreddit', 'total_words'])
data_total = data_words.sort_values(by='subreddit', ascending=True)
data_total

subreddit  total_words
2                 ama        18493
3           askreddit      1160818
11  cscareerquestions       326522
6           dankmemes        25725
5               funny        47841
7               memes        37814
8             nosleep        76217
10           politics       534476
9          psychology        32283
1             science       979181
4           worldnews       227072
12     writingprompts       268692
0                 wsb       243668

In [79]:
# from IPython.core.display import display, HTML
# display(HTML("<style>.container { width:100% !important; }</style>"))
# total_list

In [80]:
data_total['total_comments'] = num_comments
data_total['words_per_comment'] = data_total['total_words'] / data_total['total_comments']
data_total

subreddit  total_words  total_comments  words_per_comment
2                 ama        18493            1885           9.810610
3           askreddit      1160818          101141          11.477225
11  cscareerquestions       326522           12703          25.704322
6           dankmemes        25725            4077           6.309787
5               funny        47841            6659           7.184412
7               memes        37814            7332           5.157392
8             nosleep        76217            6683          11.404609
10           politics       534476           38821          13.767703
9          psychology        32283            1482          21.783401
1             science       979181           59966          16.328936
4           worldnews       227072           15265          14.875336
12     writingprompts       268692            4026          66.739195
0                 wsb       243668           36719           6.636019

In [73]:
num_dict = {'wsb': 36719, 'science': 59966, 'memes': 7332, 'cscareerquestions': 12703, 'politics': 38821, 'psychology': 1482, 'ama': 1885, 'dankmemes': 4077, 'worldnews': 15265, 'funny': 6659, 'nosleep': 6683, 'askreddit': 101141, 'writingprompts': 4026}
num_comments = [1885, 101141, 12703, 4077, 6659, 7332, 6683, 38821, 1482, 59966, 15265, 4026, 36719]

# Let's add some columns to our dataframe
data_words['total_words'] = total_list
data_words['num_comments'] = num_comments
data_words['words_per_comment'] = data_words['total_words'] / data_words['num_comments']

# Sort the dataframe by words per minute to see who talks the slowest and fastest
data_wpc_sort = data_words.sort_values(by='words_per_comment', ascending=False)
data_wpc_sort

subreddit  total_words  num_comments  words_per_comment
3           askreddit      1160818          4077         284.723571
0                 wsb       243668          1885         129.266844
11  cscareerquestions       326522          4026          81.103328
8             nosleep        76217          1482          51.428475
10           politics       534476         15265          35.013167
4           worldnews       227072          6659          34.100015
1             science       979181        101141           9.681346
12     writingprompts       268692         36719           7.317520
5               funny        47841          7332           6.524959
6           dankmemes        25725          6683           3.849319
2                 ama        18493         12703           1.455798
7               memes        37814         38821           0.974060
9          psychology        32283         59966           0.538355

CS career questions has the most words per comment on average, at 25.7. This makes sense, because people generally want to ask detailed questions and receive similarly detailed answers. Psychology has the next highest at 21, indicating lively discussion around topics. From there, words per comment drops to 14 for worldnews - people probably give relatively short replies based on their thoughts about current events. Politics at 13 is the same type of thing.

Ask reddit has 11 wpc, which makes sense based on the gist of the subreddit - a low effort, frontpage subreddit where people try to get karma, and likely give canned, generic answers in hope of getting some upvotes.

Nosleep has 11 wpc as well - as replies to fictional horror stories, this makes sense, with comments probably ranging from "good job!" to, "good job, I'll be scared to turn the lights off tonight".

AMA clocks in at 9 wpc, which also makes sense - the longer questions that span paragraphs average out with the likely large number of "lol" and other short answers.

Dankmemes has 6wpc which sounds just right - people responding to memes, but on dankmemes, so with a bit more of a personal touch - with memes coming in just below that at 5.

Finally, science has 2.9 words per comment, which seems odd at first, but makes total sense after I realized that it's probably picking up lots of " \[removed\] " comments, which really drives down the average. I have seen threads on r/science where the entire comment section is \[removed\] all the way down (with tons of nested \[removed\]s). It is a heavily moderated subreddit. I'll have to go back and skip \[removed\] and \[deleted\] comments.